pkg/commands/chgm/chgm.go

```diff
@@ -0,0 +1,137 @@
+	// OCM
+	GetClusterDeployment(clusterID string) (*hivev1.ClusterDeployment, error)
+	GetClusterInfo(identifier string) (*v1.Cluster, error)
+}
+
+// Chgm is an implementation of the Interface, and adds functionality above it
+// to craete Chgm you can use a mock, or fill it with IsRunningImpl
+type Chgm struct {
+	ChgmService
+}
+
+// isUserAllowedToStop will verify if a user is allowed to stop instances.
+// as this is the thing that might change the most it is at the top.
+// additionally it is private as I don't see anyone using this outside of AreAllInstancesRunning
+func isUserAllowedToStop(username, infraID string) bool {
+	//TODO: what is the best ordering for this? (from the most common to the most rare)
+
+	// pulled by:
+	// 1. logging (via osdctl account cli) into an aws cluster
+	// 2. running the command "aws iam list-users --query 'Users[?starts_with(UserName,`<INFRA_ID>`)].UserName'"
+	// 3. trimming the infra id from the front and the uuid from the back
+	operatorIamNames := []string{
+		"cloud-credential-operator-iam-ro",
+		"cloud-ingress-operator-credentials-aw",
+		"instance",
+		"managed-velero-operator-iam-credentia",
+		"openshift-image-registry",
+		"openshift-ingress",
+		"openshift-machine-api-aws",
+		"sre-ebs-iops-reporter-aws-credentials",
+		"sre-stuck-ebs-vols-aws-credentials",
+		"velero-iam-credentials",
+	}
+	for _, operatorIamName := range operatorIamNames {
+		// I added an additional '-' in the end to make sure it's the same as close as I can to the aws username (cannot guess the uuid)
+		operatorIamName = fmt.Sprintf("%s-%s-", infraID, operatorIamName)
+		if strings.HasPrefix(username, operatorIamName) {
+			return true
+		}
+	}
+
+	switch {
+	// all of these following cases didn't have any users connected to them
+	//case strings.HasPrefix(username, "RH-SRE"):
+	//case strings.HasPrefix(username, "osdAdmin-"):
+	//case strings.HasPrefix(username, "ManagedOpenShift-Support-"):
+	// this one on the other hand did :)
+	case strings.HasPrefix(username, "osdManagedAdmin-"):
+		return true
+	default:
+		return false
+	}
+}
+
+// AreAllInstancesRunning will check all of the instances under the externalID are running and return false if any of them are not
+func (i Chgm) AreAllInstancesRunning(externalID string) (bool, error) {
+	cluster, err := i.GetClusterInfo(externalID)
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve cluster info for %s: %w", externalID, err)
+	}
+	id := cluster.ID()
+
+	cd, err := i.GetClusterDeployment(id)
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve Cluster Deployment for %s: %w", id, err)
+	}
+	infraID := cd.Spec.ClusterMetadata.InfraID
+
+	// TODO: we are pulling ListRunningInstances and ListStoppedInstances, if we want to save on API calls, we can call ListInstances and filter from there
+	// nodeCount is only pulled here as only here we are verifying the amount of running instances
+	nodeCount := cluster.Nodes().Total()
```

Comment:
can we really rely on that the information is up to date? Are cluster auto-scaling actions updating this value? 
Otherwise, is maybe answering the following question enough for us at that point?:

_" was any master or infra node been scaled down by a not authorized user "_

Meaning someone else touched SRE managed control-plane components and thus SLA is broken. Which doesn't mean we should gather as much more information as we can.

Reply:
but a mistake by the customer done once would always appear in cloud trail, and CAD would always assign the blame to the customer for all subsequent CHGM... not easy to get this right

Reply:
we might need to take the stopped instance, search for a matching event in the timeframe to correlate to it

Reply:
You might want to modify the function doc strings. As far as I understand your function it does more than just checking if the instances are running, correct?  You also check in the end if the user is allowed to stop them :thinking:

Reply:
is that good enough?

Reply:
@ramonbutter @shibumi feel free to resolve the review thread if you feel your comments have been addressed / codified.
if I didn't. please provide concrete things you would like for me to change / card out to another time

Reply:
@georgettica I think we talked about this but I forgot to note it down. Since we now have the retry in getting the cloudtrail and we know if instances are stopped we dont need to look at the node count total. At least we can move this into a seperate PR discussion later.